src/refda/exec.c

```diff
@@ -373,93 +375,434 @@ void *refda_exec_worker(void *arg)
         }
-        else
+    }
+    else
+    {
+        CACE_LOG_DEBUG("waiting for exec event");
+        sem_wait(&(agent->execs_sem));
+    }
+
+    // execs queue may still be empty if deferred callbacks were run
+    if (refda_msgdata_queue_pop(&item, agent->execs))
+    {
+        // sentinel for end-of-input
+        const bool at_end = cace_ari_is_undefined(&(item.value));
+        if (!at_end)
         {
-            CACE_LOG_DEBUG("waiting for exec event");
-            sem_wait(&(agent->execs_sem));
+            refda_exec_exp_execset(agent, &item);
         }
+        refda_msgdata_deinit(&item);
+        if (at_end && refda_timeline_empty_p(agent->exec_timeline))
+        {
+            CACE_LOG_INFO("Got undefined exec, stopping");
+
+            // flush the input queue but keep the daemon running
+            refda_msgdata_t undef;
+            refda_msgdata_init(&undef);
+            refda_msgdata_queue_push_move(agent->rptgs, &undef);
+            sem_post(&(agent->rptgs_sem));
+
+            return false;
+        }
+    }
 
-        // execs queue may still be empty if deferred callbacks were run
-        if (refda_msgdata_queue_pop(&item, agent->execs))
+    // execute any waiting sequences
+    refda_exec_waiting(agent);
+    return true;
+}
+
+static int refda_exec_schedule_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr, bool starting);
+
+/** Execute a rule's action that has already been verified
+ * Based on code from refda_exec_exp_execset
+ */
+static int refda_exec_rule_action(refda_agent_t *agent, refda_exec_seq_t *seq, const cace_ari_t *action)
+{
+    refda_runctx_ptr_t ctxptr;
+    refda_runctx_ptr_init_new(ctxptr);
+
+    refda_runctx_t *runctx = refda_runctx_ptr_ref(ctxptr);
+
+    if (refda_runctx_from(runctx, agent, NULL))
+    {
+        return 2;
+    }
+
+    refda_runctx_ptr_set(seq->runctx, ctxptr);
+    int res = refda_exec_exp_mac(runctx, seq, action);
+
+    refda_runctx_ptr_clear(ctxptr); // Clean up extra reference created by ptr_ref
+    return res;
+}
+
+/** Begin a single execution of a time based rule
+ */
+static void refda_exec_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr)
+{
+    CHKERR1(agent);
+    CHKERR1(tbr);
+
+    if (!tbr->enabled)
+    {
+        CACE_LOG_INFO("TBR %p is not enabled", tbr);
+        return;
+    }
+
+    if (refda_amm_tbr_desc_reached_max_exec_count(tbr))
+    {
+        CACE_LOG_INFO("TBR %p reached maximum execution count", tbr);
+        refda_exec_tbr_disable(agent, tbr);
+        return;
+    }
+
+    refda_exec_seq_t *seq = refda_exec_seq_list_push_back_new(agent->exec_state);
+    seq->pid              = agent->exec_next_pid++;
+
+    // Expand rule and create exec items, CTRLs are run later by exec worker
+    if (!refda_exec_rule_action(agent, seq, &(tbr->action)))
+    {
+        tbr->exec_count++;
+        atomic_fetch_add(&agent->instr.num_tbrs_trig, 1);
+
+        // Schedule next execution of the rule now so time period is accurate
```

Comment:
Same as the SBR comment, schedule the next evaluation at the start of this function so that the interval is consistent and independent of the runtime for this function.

Reply:
Updated code accordingly. Note the actual execution takes place later on. Still, this makes scheduling independent of macro expansion and CTRL placement on the timeline.